rust/arrow/src/compute/kernels/regexp.rs

```diff
@@ -0,0 +1,117 @@
+    re: &Regex,
+    idx: usize,
+) -> Result<ArrayRef> {
+    let mut builder: GenericStringBuilder<OffsetSize> = GenericStringBuilder::new(0);
+
+    for maybe_value in array.iter() {
+        match maybe_value {
+            Some(value) => match re.captures(value) {
+                Some(caps) => {
+                    let m = caps.get(idx).ok_or_else(|| {
+                        ArrowError::ComputeError(format!(
+                            "Regexp has no group with index {}",
+                            idx
+                        ))
+                    })?;
+                    builder.append_value(m.as_str())?
+                }
+                None => builder.append_null()?,
+            },
+            None => builder.append_null()?,
+        }
+    }
+    Ok(Arc::new(builder.finish()))
+}
+
+/// Extracts a specific group matched by a regular expression for a given String array.
+/// Group index 0 returns the whole match, index 1 returns the first group and so on. Please
+/// refer to regex crate for details on pattern specifics.
+pub fn regexp_extract(array: &Array, pattern: &str, idx: usize) -> Result<ArrayRef> {
+    let re = Regex::new(pattern).map_err(|e| {
+        ArrowError::ComputeError(format!("Regular expression did not compile: {:?}", e))
+    })?;
+    match array.data_type() {
+        DataType::LargeUtf8 => generic_regexp_extract(
+            array
+                .as_any()
+                .downcast_ref::<LargeStringArray>()
+                .expect("A large string is expected"),
+            &re,
+            idx,
+        ),
+        DataType::Utf8 => generic_regexp_extract(
+            array
+                .as_any()
+                .downcast_ref::<StringArray>()
+                .expect("A string is expected"),
+            &re,
+            idx,
+        ),
+        _ => Err(ArrowError::ComputeError(format!(
+            "regexp_extract does not support type {:?}",
+            array.data_type()
+        ))),
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn extract_single_group() -> Result<()> {
+        let values = vec!["abc-005-def", "X-7-5", "X545"];
+        let array = StringArray::from(values);
+        let pattern = r".*-(\d*)-.*";
+        let actual = regexp_extract(&array, pattern, 1)?;
+        let expected = StringArray::from(vec![Some("005"), Some("7"), None]);
+        let result = actual.as_any().downcast_ref::<StringArray>().unwrap();
+        assert_eq!(&expected, result);
+        Ok(())
+    }
+
+    #[test]
+    fn no_matches() -> Result<()> {
+        let values = vec!["abc", "X::50::00", "X545"];
+        let array = StringArray::from(values);
+        let pattern = r".*-(\d*)-.*";
+        let actual = regexp_extract(&array, pattern, 1)?;
+        let expected = StringArray::from(vec![None, None, None]);
+        let result = actual.as_any().downcast_ref::<StringArray>().unwrap();
+        assert_eq!(&expected, result);
+        Ok(())
+    }
```

Comment:
I would be fine not adding this test. IMO this is covered on the test above.

Reply:
Done